src/middlewared/middlewared/alert/source/snapshot_count.py

```diff
@@ -43,11 +42,28 @@ async def check(self):
                 key=None,
             )
 
-        for dataset in sorted(datasets.keys()):
-            count = datasets[dataset]
-            if count > max_:
-                return Alert(
-                    SnapshotCountAlertClass,
-                    {"dataset": dataset, "count": count, "max": max_},
-                    key=None,
-                )
+    async def _check_smb(self) -> list[Alert]:
+        """Return an `Alert` for every dataset shared over smb whose number of snapshots exceeds the limit."""
+        max_ = await self.middleware.call("pool.snapshottask.max_count")
+        datasets = await self.middleware.call("zfs.snapshot.count")
+        smb_shares = await self.middleware.call("sharing.smb.query")
+        smb_paths = [share["path"].removeprefix("/mnt/") for share in smb_shares]
+        to_alert = list()
+
+        for path in sorted(smb_paths):
+            if path in datasets:
+                count = datasets[path]
+                if count > max_:
+                    to_alert.append(Alert(
+                        SnapshotCountAlertClass,
+                        {"dataset": path, "count": count, "max": max_},
+                        key=path,
+                    ))
+
+        return to_alert
+
+    async def check(self):
```

Comment:
You need to call `zfs.snapshot.count` in this method and then pass it around to the other methods.